### PR TITLE
[#677] Free link resources after links are closed.

### PR DESCRIPTION
**Pull request by calohmn**
Fix for #677.

### CODE REVIEW

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 1]
```diff
@@ -210,11 +211,23 @@ public final boolean supportsCapability(final Symbol capability) {
...
      */
     protected final void closeLinks(final Handler<AsyncResult<Void>> closeHandler) {
+        closeLinks(closeHandler, true);
+    }
+
+    /**
+     * Closes this client's sender and receiver links to Hono.
+     * 
+     * @param closeHandler the handler to be notified about the outcome.
+     * @param freeLinkResources if true, link resources will be freed after the links are closed
```
Comment (sysexcontrol) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: Trailing `.` is missing.

File: core/src/main/java/org/eclipse/hono/util/LinkHelper.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * Copyright (c) 2016, 2018 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
```
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/util/LinkHelper.java]: We changed the header to a more generic form ("...Contributors to the Eclipse Foundation").
See https://www.eclipse.org/hono/community/contributing/ for details.
Please adapt to this.

File: core/src/main/java/org/eclipse/hono/util/LinkHelper.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    private LinkHelper() {
+    }
+
+    /**
+     * Frees resources concerning the given link.
+     * <p>
+     * Note: this method will become obsolete with the availability of ProtonLink#free() (see vertx-proton#83).
+     *
+     * @param protonLink link to free resources for
```
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/util/LinkHelper.java]: Maybe `The link to free...  for.`?

File: core/src/main/java/org/eclipse/hono/util/LinkHelper.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                } else if (protonLink instanceof ProtonSenderImpl) {
+                    final Method getSenderMethod = ProtonSenderImpl.class.getDeclaredMethod("sender");
+                    getSenderMethod.setAccessible(true);
+                    link = (Link) getSenderMethod.invoke(protonLink);
+                } else {
+                    throw new IllegalArgumentException("unknown ProtonLink class: " + protonLink.getClass());
+                }
+                link.free();
+                LOG.debug("freed link resources");
+            } catch (Exception e) {
```
Comment (sysexcontrol) [inline core/src/main/java/org/eclipse/hono/util/LinkHelper.java]: final Exception e

File: core/src/main/java/org/eclipse/hono/util/LinkHelper.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                    link = (Link) getReceiverMethod.invoke(protonLink);
+                } else if (protonLink instanceof ProtonSenderImpl) {
+                    final Method getSenderMethod = ProtonSenderImpl.class.getDeclaredMethod("sender");
+                    getSenderMethod.setAccessible(true);
+                    link = (Link) getSenderMethod.invoke(protonLink);
+                } else {
+                    throw new IllegalArgumentException("unknown ProtonLink class: " + protonLink.getClass());
+                }
+                link.free();
+                LOG.debug("freed link resources");
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/LinkHelper.java]: IMHO we should log this at TRACE level

Comment (calohmn) [reply core/src/main/java/org/eclipse/hono/util/LinkHelper.java]: Ok, I've changed that.

**Comment by calohmn** [comment]
@sysexcontrol I've made the corresponding changes.

**Comment by sysexcontrol** [comment]
LGTM.

**Comment by sysexcontrol** [comment]
Thanks for contributing, @calohmn !

**Pull request merged**